### ISSUE TITLE
[Singleton(Mode = SingletonMode.Listener)] doesn't work for blob trigger

### ISSUE THREAD

**Issue by jon8884** [Azure/azure-webjobs-sdk-script#1633]
Hi,

I need to process items sequentially.

I have the following:

        [Singleton(Mode = SingletonMode.Listener)]
        [FunctionName("FunctionName")]        
        public static void Run([BlobTrigger(....

It spins up multiple functions in parallel.

Thanks

**Comment by jon8884** [comment]
"queues": {"batchSize": 1} in host.json fixes

**Comment by dallancarr** [comment]
👍

**Comment by ahmelsayed** [comment]
note that this will only work if you have 1 instance running. Once you have > 1, each one will get a batchSize of 1.

**Comment by jon8884** [comment]
@ahmelsayed  fair, and that makes sense.   I got hung up thinking that "queues": {"batchSize": 1} would only work for queues and not blob triggers

**Comment by ahmelsayed** [comment]
I would not have expected it to work either. I wonder if this is just a side effect of how blobTrigger implementation is done internally. The way this works is actually through queues as well. There is a scanner that scans for blob changes and then pushes a message in a hidden queue, then that queue is consumed to actually process (run your function on) the blob. 

@mathewc would know for sure.

**Comment by cmeiklejohn** [comment]
Will `[Singleton(Mode = SingletonMode.Listener)]` guarantee only a single invocation across multiple hosts?  It looks like it's not for me, because I see other instances starting up -- and blowing up because they are connecting to EventHub with a consumer group with an advanced epoch than the other consumer from the other instance.

**Comment by mathewc** [comment]
You're using the wrong mode - you want the default SingletonMode.Function. See https://github.com/Azure/azure-webjobs-sdk/wiki/Singleton for details. 

The listener mode singleton ensures that only a single host instance is running your function, however if there are many blobs that instance can process concurrently based on your concurrency settings.

Function mode singleton ensures that only a single invocation of your function runs at a time.

**Comment by cmeiklejohn** [comment]
```
Function mode singleton ensures that only a single invocation of your function runs at a time.
```

Is this true?  I've been working with Azure support on this and I'm not seeing this behavior.  I have an EventHub triggered function with the Singleton annotation (which, defaults to SingletonMode.Function) and I'm seeing multiple invocations.  After talking to Azure support, it looks like the Singleton lock only applies per host.

**Comment by mathewc** [comment]
What type of function do you have? For this to work it has to be a precompiled annotated function, not a CSX style function.

**Comment by cmeiklejohn** [comment]
I have a C# function, and this comment from 5 days ago is the exact behavior I'm seeing: Singleton does not work with the Consumption plan: https://github.com/Azure/azure-functions-host/issues/912#issuecomment-401054147.

**Comment by mathewc** [comment]
For a precompiled C# function, the singleton attribute is just passed to the underlying WebJobs SDK so should "just work".  #912 is about making this work for CSX and other languages like Node. It also mentions the billing issue for Consumption. Currently I believe the time spent waiting for the singleton will be billed.

Anyhow if you have a repro, recommend creating a new issue with all the specific details. I expect your scenario to work, and if it isn't we can find out why from a repro.

**Comment by cmeiklejohn** [comment]
I have observed different behavior and I can try to put a repro together next week.  I was advised by Azure support to switch to the App Service plan and away from Consumption to avoid this issue.

**Comment by simeyla** [comment]
@cmeiklejohn Wow. I'm having the same issue. Seems like every step of the way I run into unexpected issues with Azure services!

I can't get Singleton to do anything useful at all on consumption plan. Didn't expect that to be the issue.

I think I will have to use a timer or something to just manually process items in a queue if I really need them to be sequential.

**Comment by cmeiklejohn** [comment]
Contrary to what the documentation seems to indicate, I experimentally observed that, when on the consumption plan, the Singleton was only enforced *per host*, which means that under high load, when you're instances to expand across multiple nodes, your singleton is only enforced once-per-node, resulting in multiple invocations.

To solve this, we used Azure Blobs, along with leases, to ensure that only one invocation for a given "application entity" was invoking at a particular time.  

cc: @sebastianburckhardt

**Comment by ivorwan** [comment]
I'm with @cmeiklejohn. The docs here https://github.com/Azure/azure-webjobs-sdk/wiki/Singleton say that internally, the SingletonMode.Listener internally uses blob leases and ensures that only one function should run across scaled out instances, but I'm observing otherwise.

I'm using azure functions v2 on consumption plan, triggered by new messages on a queue
